Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs

```diff
@@ -0,0 +1,613 @@
+                case ThemePropertyTypes.Shader:
+                    SerializedProperty shaderObjectValue = item.FindPropertyRelative("Shader");
+                    EditorGUILayout.PropertyField(shaderObjectValue, new GUIContent(name, ""), false);
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        protected bool RenderDeleteButton(int index)
+        {
+            // Create Delete button if we have an array of themes
+            if (themeDefinitions.arraySize > 1 && InspectorUIUtility.SmallButton(RemoveThemePropertyContent))
+            {
+                ClearHistoryCache(index);
+                DeleteThemeDefinition((uint)index);
+
+                serializedObject.Update();
+                EditorUtility.SetDirty(theme);
+                return true;
+            }
+
+            return false;
+        }
+
+        #endregion
+
+        #region Theme Definition Management
+
+        protected virtual void AddThemeDefinition()
+        {
+            Type defaultType = typeof(InteractableActivateTheme);
+
+            ThemeDefinition newDefinition = ThemeDefinition.GetDefaultThemeDefinition(defaultType).Value;
+            if (theme.Definitions == null)
+            {
+                theme.Definitions = new List<ThemeDefinition>();
+            }
+            theme.Definitions.Add(newDefinition);
+            theme.History.Add(new Dictionary<Type, ThemeDefinition>());
+            theme.ValidateDefinitions();
+
+            serializedObject.Update();
+            EditorUtility.SetDirty(theme);
+        }
+
+        protected void DeleteThemeDefinition(uint index)
+        {
+            if (!(theme != null && theme.Definitions != null && index < theme.Definitions.Count))
+            {
+                Debug.LogError("Cannot delete ThemeDefinition. Invalid Theme object");
+                return;
+            }
+
+            theme.Definitions.RemoveAt((int)index);
+        }
+
+        private void ChangeThemeDefinitionType(int index, Type oldType, Type newType)
+        {
+            // Save theme definition to cache
+            SaveThemeDefinitionHistory(index, oldType);
+
+            // Try to load theme from history cache
+            ThemeDefinition? definition = LoadThemeDefinitionHistory(index, newType);
+            if (definition == null)
+            {
+                // if not available, then create a new one
+                definition = ThemeDefinition.GetDefaultThemeDefinition(newType);
+            }
+
+            theme.Definitions[index] = definition.Value;
+            theme.ValidateDefinitions();
+
+            themeDefinitions.serializedObject.Update();
+            EditorUtility.SetDirty(theme);
+        }
+
+        #endregion
+
+        #region Theme Definition History
+
+        protected void ClearHistoryCache(int index)
+        {
+            if (theme == null || theme.History == null || index >= theme.History.Count)
```

Comment:
I hit a bug when modifying Interactable. To repro:

Use the following theme:
![image](https://user-images.githubusercontent.com/25975362/65271378-e5cdca80-dad1-11e9-921b-19471750a393.png)


Hit "Add Theme Definition" Twice
Try to delete one of the theme definitions by pressing "Delete" button
I get the following call stack:

ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
System.ThrowHelper.ThrowArgumentOutOfRangeException (System.ExceptionArgument argument, System.ExceptionResource resource) (at :0)
System.ThrowHelper.ThrowArgumentOutOfRangeException () (at :0)
System.Collections.Generic.List`1[T].RemoveAt (System.Int32 index) (at :0)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.ClearHistoryCache (System.Int32 index) (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:501)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.RenderDeleteButton (System.Int32 index) (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:430)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.RenderThemeDefinitions () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:121)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.RenderTheme () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:70)
Microsoft.MixedReality.Toolkit.UI.Editor.ThemeInspector.OnInspectorGUI () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/VisualThemes/ThemeInspector.cs:53)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector.RenderProfileSettings () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:173)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector.RenderCustomInspector () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:94)
Microsoft.MixedReality.Toolkit.UI.Editor.InteractableInspector.OnInspectorGUI () (at Assets/MixedRealityToolkit.SDK/Inspectors/UX/Interactable/InteractableInspector.cs:83)
UnityEditor.InspectorWindow.DoOnInspectorGUI (System.Boolean rebuildOptimizedGUIBlock, UnityEditor.Editor editor, System.Boolean wasVisible, UnityEngine.Rect& contentRect) (at C:/buildslave/unity/build/Editor/Mono/Inspector/InspectorWindow.cs:1625)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)

Deleting the second to last theme definition worked, but deleting the very last theme definition did not work, probably an easy fix then.

Reply:
This has been fixed. Issue with the if condition not checking >=